src/java.base/share/classes/jdk/internal/classfile/impl/AbstractInstruction.java

```diff
@@ -0,0 +1,1308 @@
+import jdk.internal.classfile.instruction.StoreInstruction;
+import jdk.internal.classfile.instruction.TableSwitchInstruction;
+import jdk.internal.classfile.instruction.ThrowInstruction;
+import jdk.internal.classfile.instruction.TypeCheckInstruction;
+import jdk.internal.classfile.Label;
+import jdk.internal.classfile.Opcode;
+import jdk.internal.classfile.TypeKind;
+
+
+/**
+ * AbstractInstruction.
+ */
+public abstract sealed class AbstractInstruction
+        extends AbstractElement
+        implements Instruction {
+    final Opcode op;
+    final int size;
+
+    public Opcode opcode() {
+        return op;
+    }
+
+    public int sizeInBytes() {
+        return size;
+    }
+
+    public AbstractInstruction(Opcode op, int size) {
+        this.op = op;
+        this.size = size;
+    }
+
+    public abstract void writeTo(DirectCodeBuilder writer);
+
+    public static abstract sealed class BoundInstruction extends AbstractInstruction {
+        final CodeImpl code;
+        final int pos;
+
+        protected BoundInstruction(Opcode op, int size, CodeImpl code, int pos) {
+            super(op, size);
+            this.code = code;
+            this.pos = pos;
+        }
+
+        protected Label offsetToLabel(int offset) {
+            return code.getLabel(pos - code.codeStart + offset);
+        }
+
+        public void writeTo(DirectCodeBuilder writer) {
+            // Override this if the instruction has any CP references or labels!
+            code.classReader.copyBytesTo(writer.bytecodesBufWriter, pos, size);
+        }
+    }
+
+    public static final class BoundLoadInstruction
+            extends BoundInstruction implements LoadInstruction {
+
+        public BoundLoadInstruction(Opcode op, CodeImpl code, int pos) {
+            super(op, op.sizeIfFixed(), code, pos);
+        }
+
+
+        @Override
+        public TypeKind typeKind() {
+            return op.primaryTypeKind();
+        }
+
+        @Override
+        public String toString() {
+            return String.format("Load[OP=%s, slot=%d]", this.opcode(), slot());
+        }
+
+        @Override
+        public int slot() {
+            return switch (size) {
+                case 2 -> code.classReader.readU1(pos + 1);
+                case 4 -> code.classReader.readU2(pos + 2);
+                default -> throw new IllegalArgumentException("Unexpected op size: " + op.sizeIfFixed() + " -- " + op);
+            };
+        }
+
+    }
+
+    public static final class BoundStoreInstruction
+            extends BoundInstruction implements StoreInstruction {
+
+        public BoundStoreInstruction(Opcode op, CodeImpl code, int pos) {
+            super(op, op.sizeIfFixed(), code, pos);
+        }
+
+        @Override
+        public TypeKind typeKind() {
+            return op.primaryTypeKind();
+        }
+
+        @Override
+        public String toString() {
+            return String.format("Store[OP=%s, slot=%d]", this.opcode(), slot());
```

Comment:
A suggestion. I believe the `toString` output for bound and unbound instructions should be identical and it can composed solely from methods on the public instruction interface. There are some differences e.g. `Increment` and `Inc` for the unbound and bound increment instruction respectively.

If those assumptions are correct i recommend placing a static `toString` method on all unbound instructions that also have bound instructions, accepting the public instruction interface as an argument. Then the overridden `Object::toString` method defers to those. Thereby there is no duplication.
(Alas we cannot override `toString` on the instruction interface itself).

Reply:
resolved using common static format Strings